File: kotlin/provenance-mark/src/main/kotlin/com/blockchaincommons/provenancemark/DateSerialization.kt
package com.blockchaincommons.provenancemark

import com.blockchaincommons.dcbor.CborDate
import java.time.Instant
import java.time.ZoneOffset
import java.time.ZonedDateTime

private val referenceDate: Instant = ZonedDateTime
    .of(2001, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC)
    .toInstant()

object DateSerialization {
    fun serialize2Bytes(date: CborDate): ByteArray {
        val components = ZonedDateTime.ofInstant(date.instant, ZoneOffset.UTC)
        val year = components.year
        val month = components.monthValue
        val day = components.dayOfMonth

        val yy = year - 2023
        if (yy !in 0 until 128) {
            throw ProvenanceMarkException.YearOutOfRange(year)
        }
        if (month !in 1..12 || day !in 1..31) {
            throw ProvenanceMarkException.InvalidMonthOrDay(year, month, day)
        }

        val value = ((yy shl 9) or (month shl 5) or day) and 0xFFFF
        return byteArrayOf(((value shr 8) and 0xFF).toByte(), (value and 0xFF).toByte())
    }

    fun deserialize2Bytes(bytes: ByteArray): CborDate {
        require(bytes.size == 2) { "2-byte date requires exactly 2 bytes" }
        val value = ((bytes[0].toInt() and 0xFF) shl 8) or (bytes[1].toInt() and 0xFF)
        val day = value and 0b11111
        val month = (value shr 5) and 0b1111
        val yy = (value shr 9) and 0b1111111
        val year = yy + 2023

        if (month !in 1..12 || day !in rangeOfDaysInMonth(year, month)) {
            throw ProvenanceMarkException.InvalidMonthOrDay(year, month, day)
        }

        return try {
            CborDate.fromYmdHms(year, month, day, 0, 0, 0)
        } catch (_: Exception) {
            throw ProvenanceMarkException.InvalidDate(
                "Cannot construct date $year-${"%02d".format(month)}-${"%02d".format(day)}"
            )
        }
    }

    fun serialize4Bytes(date: CborDate): ByteArray {
        val seconds = date.instant.epochSecond - referenceDate.epochSecond
        if (seconds < 0 || seconds > UInt.MAX_VALUE.toLong()) {
            throw ProvenanceMarkException.DateOutOfRange("seconds value too large for u32")
        }
        val n = seconds.toUInt()
        return byteArrayOf(
            (n shr 24).toByte(),
            (n shr 16).toByte(),
            (n shr 8).toByte(),
            n.toByte(),
        )
    }

    fun deserialize4Bytes(bytes: ByteArray): CborDate {
        require(bytes.size == 4) { "4-byte date requires exactly 4 bytes" }
        val n =
            ((bytes[0].toUInt() and 0xFFu) shl 24) or
                ((bytes[1].toUInt() and 0xFFu) shl 16) or
                ((bytes[2].toUInt() and 0xFFu) shl 8) or
                (bytes[3].toUInt() and 0xFFu)
        val instant = referenceDate.plusSeconds(n.toLong())
        return CborDate.fromInstant(instant)
    }

    fun serialize6Bytes(date: CborDate): ByteArray {
        val millis = date.instant.toEpochMilli() - referenceDate.toEpochMilli()
        if (millis < 0) {
            throw ProvenanceMarkException.DateOutOfRange("milliseconds value too large for u64")
        }
        val n = millis.toULong()
        if (n > 0xe5940a78a7ffuL) {
            throw ProvenanceMarkException.DateOutOfRange("date exceeds maximum representable value")
        }
        val full = ByteArray(8)
        var value = n
        for (i in 7 downTo 0) {
            full[i] = (value and 0xFFu).toByte()
            value = value shr 8
        }
        return full.copyOfRange(2, 8)
    }

    fun deserialize6Bytes(bytes: ByteArray): CborDate {
        require(bytes.size == 6) { "6-byte date requires exactly 6 bytes" }
        val full = ByteArray(8)
        System.arraycopy(bytes, 0, full, 2, 6)
        var n = 0uL
        for (b in full) {
            n = (n shl 8) or (b.toUByte().toULong())
        }
        if (n > 0xe5940a78a7ffuL) {
            throw ProvenanceMarkException.DateOutOfRange("date exceeds maximum representable value")
        }
        val instant = referenceDate.plusMillis(n.toLong())
        return CborDate.fromInstant(instant)
    }
}

fun CborDate.serialize2Bytes(): ByteArray = DateSerialization.serialize2Bytes(this)
fun CborDate.serialize4Bytes(): ByteArray = DateSerialization.serialize4Bytes(this)
fun CborDate.serialize6Bytes(): ByteArray = DateSerialization.serialize6Bytes(this)

fun deserialize2Bytes(bytes: ByteArray): CborDate = DateSerialization.deserialize2Bytes(bytes)
fun deserialize4Bytes(bytes: ByteArray): CborDate = DateSerialization.deserialize4Bytes(bytes)
fun deserialize6Bytes(bytes: ByteArray): CborDate = DateSerialization.deserialize6Bytes(bytes)

fun rangeOfDaysInMonth(year: Int, month: Int): IntRange {
    val nextMonth = if (month == 12) {
        ZonedDateTime.of(year + 1, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC)
    } else {
        ZonedDateTime.of(year, month + 1, 1, 0, 0, 0, 0, ZoneOffset.UTC)
    }
    val lastDay = nextMonth.minusDays(1).dayOfMonth
    return 1..lastDay
}

File: csharp/ProvenanceMark/ProvenanceMark/DateSerialization.cs
using System.Buffers.Binary;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.ProvenanceMark;

/// <summary>
/// Resolution-dependent compact date serialization helpers.
/// </summary>
public static class DateSerialization
{
    private static readonly DateTimeOffset ReferenceDate =
        new(2001, 1, 1, 0, 0, 0, TimeSpan.Zero);

    private const ulong MaxSixByteMilliseconds = 0xe5940a78a7ffUL;

    public static byte[] Serialize2Bytes(CborDate date)
    {
        var utc = date.DateTimeValue.ToUniversalTime();
        var year = utc.Year;
        var month = utc.Month;
        var day = utc.Day;

        var yearOffset = year - 2023;
        if (yearOffset is < 0 or >= 128)
        {
            throw ProvenanceMarkException.YearOutOfRange(year);
        }

        if (month is < 1 or > 12 || day is < 1 or > 31)
        {
            throw ProvenanceMarkException.InvalidMonthOrDay(year, month, day);
        }

        var value = ((yearOffset << 9) | (month << 5) | day) & 0xffff;
        return
        [
            (byte)((value >> 8) & 0xff),
            (byte)(value & 0xff)
        ];
    }

    public static CborDate Deserialize2Bytes(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length != 2)
        {
            throw new ArgumentException("2-byte date requires exactly 2 bytes", nameof(bytes));
        }

        var value = (bytes[0] << 8) | bytes[1];
        var day = value & 0b11111;
        var month = (value >> 5) & 0b1111;
        var year = ((value >> 9) & 0b1111111) + 2023;

        if (month is < 1 or > 12 || day < 1 || day > RangeOfDaysInMonth(year, month))
        {
            throw ProvenanceMarkException.InvalidMonthOrDay(year, month, day);
        }

        try
        {
            return CborDate.FromYmdHms(year, month, day, 0, 0, 0);
        }
        catch (Exception)
        {
            throw ProvenanceMarkException.InvalidDate(
                $"Cannot construct date {year:D4}-{month:D2}-{day:D2}");
        }
    }

    public static byte[] Serialize4Bytes(CborDate date)
    {
        var seconds = checked((long)(date.DateTimeValue.ToUniversalTime() - ReferenceDate).TotalSeconds);
        if (seconds < 0 || seconds > uint.MaxValue)
        {
            throw ProvenanceMarkException.DateOutOfRange("seconds value too large for u32");
        }

        var data = new byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(data, (uint)seconds);
        return data;
    }

    public static CborDate Deserialize4Bytes(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length != 4)
        {
            throw new ArgumentException("4-byte date requires exactly 4 bytes", nameof(bytes));
        }

        var seconds = BinaryPrimitives.ReadUInt32BigEndian(bytes);
        return CborDate.FromDateTime(ReferenceDate.AddSeconds(seconds));
    }

    public static byte[] Serialize6Bytes(CborDate date)
    {
        var utc = date.DateTimeValue.ToUniversalTime();
        var milliseconds = utc.ToUnixTimeMilliseconds() - ReferenceDate.ToUnixTimeMilliseconds();
        if (milliseconds < 0)
        {
            throw ProvenanceMarkException.DateOutOfRange("milliseconds value too large for u64");
        }

        var value = (ulong)milliseconds;
        if (value > MaxSixByteMilliseconds)
        {
            throw ProvenanceMarkException.DateOutOfRange("date exceeds maximum representable value");
        }

        var full = new byte[8];
        BinaryPrimitives.WriteUInt64BigEndian(full, value);
        return full[2..];
    }

    public static CborDate Deserialize6Bytes(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length != 6)
        {
            throw new ArgumentException("6-byte date requires exactly 6 bytes", nameof(bytes));
        }

        var full = new byte[8];
        bytes.CopyTo(full.AsSpan(2));
        var value = BinaryPrimitives.ReadUInt64BigEndian(full);
        if (value > MaxSixByteMilliseconds)
        {
            throw ProvenanceMarkException.DateOutOfRange("date exceeds maximum representable value");
        }

        return CborDate.FromDateTime(ReferenceDate.AddMilliseconds(value));
    }

    public static int RangeOfDaysInMonth(int year, int month)
    {
        return DateTime.DaysInMonth(year, month);
    }
}
